Add support for using standard headers as modules.

worthless443 opened the pull request:
As of the current commit, I can not see if it is possible to import standard headers as modules, for example, in `example/hello.cc`:
```c++
import <cstdio>;
```
or in `example/main.cc` if you want to use headers as modules, namely:
```c++
import hello;
import <iostream>
```
I'm not sure if there's already support for this as of current, but for me, it doesn't work and I want to use everything with the `import` syntax when modules are concerned and yes modern C++ too ;) 

so `example/hello.cc` looks like:
```c++
module;

import <cstdio>;

export module hello;

export void hello() { std::printf("Hello, modules!\n"); }
```
and `example/main.cc`:
```c++
import hello;
import <iostream>;
int main() { hello(); std::cout << "hello from standard as modules!\n"; }
```
the change `add_executable()` to `add_stdmodular_executable()` is necessary when:
 - your main file contains `import`s of standard headers 
 - or you are only compiling the main file (solely)
 
 (it doesn't work with Clang yet, therefore emits errors if in any of your files contains imports of std headers and CXX compiler is Clang).
(the code is long and a bit messed up, the names are weird, so please suggest me the names)

worthless443 committed:
* added support for using standard headers as modules
* if used the import syntax with standard headers, and the compiler is
  Clang, cmake will fail (don't yet know how to do it with Clang)
* added a CMake command that shall be used either when using std
  headers in the main executable or want to use std headers as
  executables and compile it soley (without other moduler libraries)

worthless443 committed:
use moduler import syntax

worthless443 committed:
typo

Mixaill commented:
@worthless443 Thank you for your contribution!

@vitaut In general, do we want to have a self-made C++23 functionality inside this library? Maybe we should do it as a "plugin"?

```
include(modules.cmake)
include(modules_extra.cmake)
```

worthless443 committed:
version check, and fail while configuring if GCC version <12 and has used import syntax for std headers

worthless443 commented:
It seems like the gcm s are built after compiling the imports.

vitaut commented:
> In general, do we want to have a self-made C++23 functionality inside this library?

I think it's fine provided that the functionality is an opt-in. We don't need to split the CMake "library", having this exposed as separate functions or options to existing functions seems OK.

vitaut commented on example/CMakeLists.txt:
```diff
@@ -24,5 +24,5 @@ if(cxx_modules)
   set(app_src main_fallback.cc)
 endif()
-add_executable(main ${app_src})
+add_stdmodular_executable(main ${app_src})
```
I think it better be a flag for `add_executable`, e.g. `STD_HEADER_UNITS`.

vitaut replied:
Nevermind, `add_executable` is a standard function.

worthless443 replied:
yes, and also I didn't find any way to overload/replace `add_executable` that lets you use `add_executable` having a parameter when set to, for example, `STD_HEADER_UNITS` or or `STD_HEADER_IMPORTS` will use a wrapper around the function by the same name (but with different arguments), but by default CMake only sees the overloaded function and the original function with the same name, and will result in a recursive call. 

It could have been possible to do away with some tricks to `cmake_parse_arguments()`, but I don't know how.

Can you suggest a better for `add_stdmodular_executable`?

vitaut replied:
Having a separate function makes sense. I suggest renaming it to `add_module_executable` for consistency with `add_module_library`.

worthless443 committed:
add dependencies so the gcms are built before getting to the main targets(?)

worthless443 committed:
forgot to add the gcm dependencies for add_module_library(), hopefully it works by now

worthless443 commented:
@vitaut why are tests with GCC on linux marked as cancelled? As of the commit 226006638198f6531e57ed22a965b62805e8af16, is mentioned that Clang will not work with having imports of standard headers, and is only limited to GCC. Can you suggest a way how to make do with getting it working with Clang?

Mixaill commented:
>why are tests with GCC on linux marked as cancelled? 

In case of an error, github cancels unrun tests. If you want to focus on GCC, you can temporarily remove Clang here
https://github.com/vitaut/modules/blob/master/.github/workflows/linux.yml#L16

> Can you suggest a way how to make do with getting it working with Clang?

For non-GCC compilers you can use something like `FALLBACK`, you can look at `add_module_library()` for example.

worthless443 commented:
Thanks for the response @Mixaill ! I don't particularly want to focus on GCC though, it's the thing that I can't get Clang to work with compiling sources with standard header gcms built, Clang does build them, and having used all the flags complaint to C++20 modules, and adding the directory containing the gcms built as a flag to Clang, I still can't get it to work. I do have the latest Clang, however not sure how to make do with it. 
[ I don't know a thing about MSVC, so I decided to completely leave that out ]

>  you can use something like FALLBACK

yes, sure but that's not really that cool. Firstly, IMO,  `add_module_library` should fail if the compiler doesn't support C++20 modules and not rely on `FALLBACK` and silently fail (it should fail loudly). Secondly, the user might not have FALLBACK sources set up at all, in which case it will not be useful.

vitaut commented on example/hello.cc:
```diff
@@ -1,6 +1,6 @@
 module;
 
-#include <cstdio>
+import <cstdio>;
```
Let's keep #include here and introduce another example/test that uses imports. The new example can be only tested on supported compiler(s).

worthless443 committed:
* added explict flag for catching any configuration or compiler-related error
* added a flag for forcing not building standard modules gcm (useless but could be necessary)
* change functino name `add_stdmodular_executable` to `add_module_executable` with extra flags for both `add_module_executable` or `add_module_library`.
* fail if any of the sources include the syntax `import <stdheader>;` and the C++ compiler is Clang (the CATCH_ERROR flag can be set to TRUE if want to fail but by default, it is set to Warn).
* As it automatically detects if the sources have standard imports, an explicit `FORCE_NO_MODULE_STD` can be set to `TRUE` to force it to not compile additioal targets required for doing so

[Note: the error or warning when using Clang only appears if any of your sources contain imports of std headers]

worthless443 committed:
only GCC 12

worthless443 committed:
for now, only C++20

worthless443 commented:
the CI should not fail now. Removed windows (not interested).

vitaut commented:
I think this needs a lot more work before it is ready. I recommend experimenting with it in a fork and resubmitting once it is working and doesn't affect existing functionality.
